include/aws/s3/private/s3_meta_request_impl.h

```diff
@@ -45,26 +46,32 @@ typedef void(aws_s3_meta_request_prepare_request_callback_fn)(
     void *user_data);
 
 struct aws_s3_prepare_request_payload {
+    struct aws_allocator *allocator;
     struct aws_s3_request *request;
     aws_s3_meta_request_prepare_request_callback_fn *callback;
     void *user_data;
     struct aws_task task;
+    struct aws_future *preparation_future; /* aws_future<void> this operation is waiting on */
 };
 
 struct aws_s3_meta_request_vtable {
     /* Update the meta request.  out_request is required to be non-null. Returns true if there is any work in
      * progress, false if there is not. */
     bool (*update)(struct aws_s3_meta_request *meta_request, uint32_t flags, struct aws_s3_request **out_request);
 
+    /* Run vtable->prepare_request() on the meta-request's event loop.
+     * We do this because body streaming is slow, and we don't want it on our networking threads.
+     * The callback may fire on any thread (an async sub-step may run on another thread). */
```

Comment:
more a question for later: do we have any cases where firing it on any thread is not what we want? interrupting other work on that thread? numa?

Reply:
we'll see how it shakes out in profiling

I'm currently just using the future's vanilla [register_callback()](https://github.com/awslabs/aws-c-io/blob/cbea39a7f40b1892bc4bf148234c805b7cc14192/include/aws/io/future.h#L64-L71) function, which fires the callback immediately when the future completes. I'm assuming we want to react as quickly as possible after the read completes.

We could also use the future's [register_event_loop_callback()](https://github.com/awslabs/aws-c-io/blob/cbea39a7f40b1892bc4bf148234c805b7cc14192/include/aws/io/future.h#L91-L103) function, so that the callback is always scheduled to run on the same IO thread. But if there's some other long-running thing happening on that thread, this would add delay in how quickly we can react to a slow read() call finishing up.